### ISSUE TITLE
Power Management for reboot should call new method

### ISSUE THREAD

**Issue by nvtkaszpir** [cobbler/cobbler#2366]
### Is your feature request related to a problem?

Feature request.
Currently `reboot` command actually invokes `power off` and then `power on`, but some systems allow to run reboot (or cycle) command, which is let say 'soft' power off' that does not trigger full power outage for the hardware.
See a the bottom for additional context.

Another idea is to introduce new option such as 'soft reboot'  for 'reboot' and 'hard reboot' for power cycling as it is used now.

### The Behaviour you'd like

Modify method in https://github.com/cobbler/cobbler/blob/master/cobbler/power_manager.py#L301 so it actually passes  "reboot" parameter. This way user may use custom template to fine tune the command passed down (for example certain systems prefer 'cycle' and not 'reboot'), especially if you actually have a different logic between power on/off and reset signal.

### Alternatives you've considered

Was also thinking about detecting events in short time span on the fence_* script - 'power_off and on in short time span in actually a reboot request' - but that increase complexity thus causing even more problems later on.

### Additional context

From power perspective:
Reset - (soft power cycle) just as if you pressed reset key on the computer case (or shortening pins on the motherboard). This will send reset signal to the hardware but will not trigger full power off. This usually does not cause excessive power strain on the hardware.
Power Cycle - (more like a power off, wait few seconds, power on) hold power key for over 5s to trigger power off, thus totally disconnecting power. After the while system does power on if power key is pressed. This shuts down system on power level, cutting out electricity, thus may cause additional power strain on the hardware.

**Comment by SchoolGuy** [comment]
This was to my knowledge in the past for systems which don't support a reboot. We need to check if now every system we support (or try to) is able to to that. Before that I wouldn't like to have two implementations in parallel in Cobbler. We have enough legacy code which is causing problems or complicating things.

Fun fact: Some `fence_*` implement the reboot the same way we do, so even if we change our implementation we cannot guarantee that this actually does what you propose. At least this information is the last I was aware of.

**Comment by nvtkaszpir** [comment]
Yes I'm aware this causes some issues especially for backward compatibility.
My purpose is just to support "reboot" as a passed down parameter, some fencing agents support it (and they decide what to do with it).

Current implementation actually masks the way it is implemented by fencing agents - in fact cobbler expects that agent supports only 'power off', 'power on' and 'status' commands.

That's why I was a bit surprised to discover that the reboot is not doing what it sounds - I was expecting it to pass reboot command down the fencing agents, and ten react accordingly. Instead it does its own logic which may not be compatible with the reboot implemented in the agents. In that case no wonder it will be breaking the current ways of using reboot, because it no longer will be a off/on operation, but will depend on the actions of the fencing agent. For some systems that may still be insufficient and may enforce to add something like a new operation such as 'power cycle the way cobbler did it' to actually keep old behavior.

**Comment by SchoolGuy** [comment]
Okay that is a very convincing argumentation. I will have a look when we can do this.

**Comment by watologo1** [comment]
You want to pass --reboot to fence-agents, right?

Still a kind of --reset should be kept doing --on/--off as it is currently done.

It should be done this way, otherwise naming is too confusing (new reset function ending up as fence-agents reboot...).

Do you have a concrete example where this may (urgently needed or doesn't work otherwise?) help?

Another idea/interface could/would be to directly try to ssh into the client to soft-reset:
ssh --idendity-file="mySSHkey" root@system_name reboot
This would do a proper unmounting and shutdown, etc.

**Comment by nvtkaszpir** [comment]
> You want to pass --reboot to fence-agents, right?
Yes, most agents understands `reboot`, some also alias it as `reset`.

> Still a kind of --reset should be kept doing --on/--off as it is currently done.
> It should be done this way, otherwise naming is too confusing (new reset function ending up as fence-agents reboot...).
Yeah, that's correct, it should be named after supported action in fencing agents.

> Do you have a concrete example where this may (urgently needed or doesn't work otherwise?) help?
> Another idea/interface could/would be to directly try to ssh into the client to soft-reset:
> ssh --idendity-file="mySSHkey" root@system_name reboot
> This would do a proper unmounting and shutdown, etc.
This is exact use case I have now, wrote simple fence_ssh fencing agent so that it just runs ssh command on the host to invoke reboot.

In the beginning I did very ugly thing but better not mention it here  (little sneak peak - power off is doing reboot) ;-)
Now it's more like a custom power_management.py file which does what it should (power on/off/reboot), where power on is invoking wake on LAN command.

I'm also in the process of doing simple raspberry-pi controlled relay switches for separate power on/off/hard-off/ on one switch and reset on another (probably will be backed up with node-red), with checking current power status via power led. The overall functionality is more like a custom-made BMC for multiple physical devices.

**Comment by watologo1** [comment]
> This is exact use case I have now, wrote simple fence_ssh fencing agent so that it just runs ssh command on the host to invokereboot.

I very much like the idea. But we should think this to an end and a proper implementation if this should hit mainline.
I think this should be separate options. An own identity file that can get passed via a new parameter.
More should not be needed? Host is known..., user/pass plaintext is not an option.

IMO if netboot is triggered, it should always be tried to first log in via ssh if this param/variable is/got set and try to issue a "reboot" to unmount devices, etc.

If there is a timeout or the ssh key does not exists, the existing code takes place as usual.

This should be rather compatible and riskless (beside a possible ssh timeout of some seconds and a possible "reboot gets stuck, but reported success...).

But we do this already with a similar tool (which gets replace by cobbler partly atm).

**Comment by nvtkaszpir** [comment]
I suggest to focus on implementing passing reboot command to arbitrary fence agent.
For now I would remove fence_ssh from the feature request, because there is no official code for that AFAIR.

In general cobbler has retry mechanism but AFAIR it does not allow for extended timeout control.
Some fencing agents already implement some checks to test if the command actually caused the reboot process. Yet some timeouts are hardcoded or too short in certain situations - for example some RAID arrays take ages to boot and I know cases of  people doing `power off, power on knowing it will fail, go take coffee, check power on again, continue your work`. 

fence_ssh indeed introduces additional complexity such as:
- notice that you may have a right to power on/off, provision the host, but without the right to ssh to it.
- if user/pass is not acceptable (though other fencing tools still can use it as stated in the docs), then it is needed to add public key used by cobbler instance to ssh to the managed host - and some may not be eager to allow that.
- moreover checking ssh without checking host fingerprint could lead to powering off wrong host ( DNS/DHCP cahcing and so on)
- if someone is really eager to handle ssh connection then it is possible to adjust ssh config file to handle with special cases, such as different ssh key to use and so on.
So I'd leave fence_ssh for later, or people that are eager to use it or know what they are asking for.

**Comment by SchoolGuy** [comment]
@watologo1 I think if we touch that we need to simplify the code. Currently this is in my opinion to custom. So if you touch that I would very much appreciate if we could clean this up. Otherwise we just add more custom code to handle special cases.

One thing which should be more or less trivial would be to clean up the logging for the class.

**Comment by SchoolGuy** [comment]
3.4.0 is quite big already and we are late with the planned release date. Thus let's move that to 3.5.0.
